quilt/test/test_cli.py

```diff
@@ -0,0 +1,652 @@
+        """Create a parameter tree from an argparse parser.
+
+        Once an ArgumentParser is configured, it can be passed as the param
+        `argparse_object`, which makes the parameters available as a tree.
+
+        If any parameter has an item from `ignore` in its opton strings, it
+        will be excluded from the tree and from iteration.
+
+        The `leaf_func` param warrants a bit more of an in-depth description.
+        When an object is found not to have any further children, it is passed
+        to the leaf_func.  The returned value is then entered into the tree.
+        The default function simply returns the object's "dest" attribute,
+        which acts as a very short description of what the argument is, in
+        this case.
+
+        :param argparse_object: A (configured) ArgumentParser object
+        :param ignore: Ignore arguments in given iterable (-h, --help by default)
+        :param leaf_func: Receives an object.  Data returned is entered into tree.
+        """
+        self.data = argparse_object
+        self.ignored = ignore
+        self.leaf_func = leaf_func
+
+    def __getitem__(self, k):
+        if k in self.ignored:
+            raise KeyError("Ignored: {!r}".format(k))
+
+        if isinstance(k, int):
+            positionals = self.data._get_positional_actions()
+            try:
+                action = positionals[k]
+            except IndexError:
+                raise KeyError(k)
+            choices = action.choices
+            if not choices:
+                return self.leaf_func(action)
+            return collections.OrderedDict((choice, ArgparseIntrospector(parser))
+                                            for choice, parser in choices.items())
+        else:
+            optionals = self.data._get_optional_actions()
+            for o in optionals:
+                if k in o.option_strings:
+                    return self.leaf_func(o)
+        raise KeyError(k)
+
+    def __len__(self):
+        return len(tuple(self.__iter__()))
+
+    def __iter__(self):
+        for index, positional in enumerate(self.data._get_positional_actions()):
+            yield index
+        for optional in self.data._get_optional_actions():
+            if any(s in self.ignored for s in optional.option_strings):
+                continue
+            yield optional.option_strings[0]
+
+    def as_dict(self):
+        def _mcopy(old):
+            if not old:
+                return self.leaf_func(old.data)
+            if not isinstance(old, collections.Mapping):
+                return old
+            new = {}
+            for k, v in old.items():
+                new[k] = _mcopy(v) if v else None
+            return new
+        return _mcopy(self)
+
+    def __repr__(self):
+        args = '\n'.join(repr(x) for x in self)
+        return 'ArgparseIntrospector with args:\n' + args
+
+
+class RecursiveMappingWrapper(collections.MutableMapping):
```

Comment:
This feels really over-engineered to me, to be honest...

Reply:
It was a piece of code from my personal library that turns any mapping into a recursively-accessible mapping.  Since it was more than is needed, and we have no need for it elsewhere, I've taken just the snippets we need out, and integrated them into ArgparseIntrospector now.